cranelift/frontend/src/frontend.rs

```diff
@@ -804,6 +805,130 @@ impl<'a> FunctionBuilder<'a> {
+        left: Value,
+        right: Value,
+        size: Value,
+    ) -> Value {
+        let pointer_type = config.pointer_type();
+        let signature = {
+            let mut s = Signature::new(config.default_call_conv);
+            s.params.reserve(3);
+            s.params.push(AbiParam::new(pointer_type));
+            s.params.push(AbiParam::new(pointer_type));
+            s.params.push(AbiParam::new(pointer_type));
+            s.returns.push(AbiParam::new(types::I32));
+            self.import_signature(s)
+        };
+
+        let libc_memcmp = self.import_function(ExtFuncData {
+            name: ExternalName::LibCall(LibCall::Memcmp),
+            signature,
+            colocated: false,
+        });
+
+        let call = self.ins().call(libc_memcmp, &[left, right, size]);
+        self.func.dfg.first_result(call)
+    }
+
+    /// Optimised [`Self::call_memcmp`] for small copies.
+    ///
+    /// This implements the byte slice comparison `int_cc(left[..size], right[..size])`.
+    ///
+    /// `left_align` and `right_align` are the statically-known alignments of the
+    /// `left` and `right` pointers respectively.  These are used to know whether
+    /// to mark `load`s as aligned.  It's always fine to pass `1` for these, but
+    /// passing something higher than the true alignment may trap or otherwise
+    /// misbehave as described in [`MemFlags::aligned`].
+    ///
+    /// Note that `memcmp` is a *big-endian* and *unsigned* comparison.
+    /// As such, this panics when called with `IntCC::Signed*` or `IntCC::*Overflow`.
+    pub fn emit_small_memory_compare(
+        &mut self,
+        config: TargetFrontendConfig,
+        int_cc: IntCC,
+        left: Value,
+        right: Value,
+        size: u64,
+        left_align: std::num::NonZeroU8,
+        right_align: std::num::NonZeroU8,
+        flags: MemFlags,
+    ) -> Value {
+        use IntCC::*;
+        let (zero_cc, empty_imm) = match int_cc {
+            //
+            Equal => (Equal, true),
+            NotEqual => (NotEqual, false),
+
+            UnsignedLessThan => (SignedLessThan, false),
+            UnsignedGreaterThanOrEqual => (SignedGreaterThanOrEqual, true),
+            UnsignedGreaterThan => (SignedGreaterThan, false),
+            UnsignedLessThanOrEqual => (SignedLessThanOrEqual, true),
+
+            SignedLessThan
+            | SignedGreaterThanOrEqual
+            | SignedGreaterThan
+            | SignedLessThanOrEqual => {
+                panic!("Signed comparison {} not supported by memcmp", int_cc)
+            }
+            Overflow | NotOverflow => {
+                panic!("Overflow comparison {} not supported by memcmp", int_cc)
+            }
+        };
+
+        if size == 0 {
+            return self.ins().bconst(types::B1, empty_imm);
+        }
+
+        // Future work could consider expanding this to handle more-complex scenarios.
+        fn type_from_byte_size(size: u64) -> Option<Type> {
```

Comment:
Hmm, `Type::int` is already mostly `Type::int_with_bit_size`, just with a different parameter type.

Here's a few possible directions.  Let me know which you'd prefer I do, @cfallin 

- Add `Type::int_with_byte_size` and `Type::int_with_bit_size`, concentrating on the "different integer width" use by having them take `impl TryInto<u16>`.
- Add a `Type::int_bytes` that takes `u16`
- Rename `Type::int` to `Type::int_bits` and add `Type::int_bytes` (rg only spots 10 obvious uses of `Type::int` so it might not be as horrible as I'd have guessed)
- One of the above, but using a type other than `u16` for `Type::int` (TBH `u16` is kinda odd there, since the biggest value it cares about is 128 and would fit in a `u8`, but it's also not the `u32` that rust's `std` [tends to use for numbers of bits](https://doc.rust-lang.org/nightly/std/primitive.u16.html#method.trailing_ones))
- Some combination thereof
- Make a follow-up PR for one of the above
- Do nothing

Reply:
Ah, I had missed the existing `Type::int()` -- IMHO, it might be better to leave that as-is (`Type::int(32)` is simple/ergonomic/very clear) and then add this as `Type::int_with_byte_size()` (I slightly prefer this over `int_bytes`, the latter sounds like it's a special kind of int that's a vector of bytes or something like that). And I guess let's match the `u16` for consistency (I'm not sure why `u16`, maybe to allow 256-bit vecs but still be somewhat compact?).

Thanks!

Reply:
> I'm not sure why u16, maybe to allow 256-bit vecs but still be somewhat compact?

That would make sense if it was stored, but `Type` is just `pub struct Type(u8);`, so it's just a parameter, and thus the choice seems like it could be pretty much anything.

I'll go add `Type::int_with_byte_size(u16) -> Type` as requested.

Reply:
OK, yeah, that existing bit is unclear then; but, better saved for a later cleanup :-) Thanks!